aw-datastore/src/datastore.rs

```diff
@@ -581,4 +598,84 @@ impl DatastoreInstance {
     }
+
+    pub fn insert_value(&self, conn: &Connection, key: &str, data: &str)
+        -> Result<(), DatastoreError> {
+
+        let mut stmt = match conn.prepare("
+                INSERT OR REPLACE INTO key_value(key, value)
+                VALUES (?1, ?2)") {
+            Ok(stmt) => stmt,
+            Err(err) => return Err(DatastoreError::InternalError(
+                format!("Failed to prepare insert_value SQL statement: {}", err)
+            ))
+        };
+        stmt.execute(&[key, data]).expect(
+            &format!("Failed to insert key-value pair: {}", key)
+        );
+        return Ok(())
+    }
+
+    pub fn delete_value(&self, conn: &Connection, key: &str) -> Result<(), DatastoreError>{
+        conn.execute("DELETE FROM key_value WHERE key = ?1", &[key])
+            .expect("Error deleting value from database");
+        return Ok(())
+    }
+
+    pub fn get_value(&self, conn: &Connection, key: &str) -> Result<String, DatastoreError>{
+        let mut stmt = match conn.prepare("
+                SELECT * FROM key_value WHERE KEY = ?1") {
+            Ok(stmt) => stmt,
+            Err(err) => return Err(DatastoreError::InternalError(
+                format!("Failed to prepare get_value SQL statement: {}", err)
+            ))
+        };
+
+        return match stmt.query_row(&[key], |row|{
+            row.get(1)
+        }) {
+            Ok(result)  => Ok(result),
+            Err(err) => match err {
+                rusqlite::Error::QueryReturnedNoRows => Err(DatastoreError::NoSuchValue),
+                _ => Err(DatastoreError::InternalError(
+                              format!("Get value query failed for key {}", key))
+                )
+            }
+        }
+    }
+
+    pub fn get_values_starting(&self, conn: &Connection, pattern: &str) 
+        -> Result<HashMap<String, String>, DatastoreError>
+    {
+        let mut stmt = match conn.prepare("SELECT * FROM key_value WHERE key LIKE '.%' ") {
+            Ok(stmt) => stmt,
+            Err(err) => return Err(DatastoreError::InternalError(
+                format!("Failed to prepare get_value SQL statement: {}", err)
+            ))
+        };
+
+        let mut output = HashMap::<String, String>::new();
+        // Unwrap to String tuple or panic on SQL row if types are invalid.
+        // should never happen with a properly initialized table
+        let result = stmt.query_map(&[pattern], |row| Ok(
+            (row.get::<usize, String>(0).expect("Invalid sql column types"), 
+             row.get::<usize, String>(1).expect("Invalid sql column types")))
+        );
+
+        match result {
+            Ok(kv_pairs) => {
+                for row in kv_pairs {
+                    let (k, v) = row.unwrap();
+                    output.insert(k, v);
+                }
+                Ok(output)
+            },
+            Err(err) => match err {
+                rusqlite::Error::QueryReturnedNoRows => Err(DatastoreError::NoSuchValue),
```

Comment:
Shouldn't it be "NoSuchKey" rather than "NoSuchValue"?
A key should always have a value so "NoSuchValue" should be impossible.

Reply:
That does make more sense. Will change.

Reply:
Fixed in 8e07f60